plugins/cloud-azure/src/main/java/org/elasticsearch/cloud/azure/storage/AzureStorageServiceImpl.java

```diff
@@ -175,12 +175,14 @@ public OutputStream getOutputStream(String container, String blob) throws URISyn
 
         CloudBlobContainer blob_container = client.getContainerReference(container);
         if (blob_container.exists()) {
-            for (ListBlobItem blobItem : blob_container.listBlobs(keyPath + prefix)) {
+            for (ListBlobItem blobItem : blob_container.listBlobs(keyPath + (prefix == null ? "" : prefix))) {
                 URI uri = blobItem.getUri();
                 logger.trace("blob url [{}]", uri);
-                String blobpath = uri.getPath().substring(container.length() + 1);
-                BlobProperties properties = blob_container.getBlockBlobReference(blobpath).getProperties();
-                String name = blobpath.substring(keyPath.length() + 1);
+                String blobPath = uri.getPath().substring(1 + container.length() + 1);
```

Comment:
Yes, it's intentional. The `uri.getPath()` returns something that looks like `/container/keyPath`. Taking `uri.getPath().substring(container.length() + 1)` gives `/keyPath` and that leading `/` confuses Azure because it ultimately gets appended to the storage URI. An extra `1 + ` removes that additional `/` so that the correct URI is constructed.

I will add a comment. Thanks @tlrx!